components/canvas/webgl_thread.rs

```diff
@@ -0,0 +1,1191 @@
+
+/// A WebGLThread manages the life cycle and message multiplexing of
+/// a set of WebGLContexts living in the same thread.
+pub struct WebGLThread<VR: WebVRRenderHandler + 'static, OB: WebGLThreadObserver> {
+    /// Factory used to create a new GLContext shared with the WR/Main thread.
+    gl_factory: Arc<GLContextFactory>,
+    /// Channel used to generate/update or delete `webrender_api::ImageKey`s.
+    webrender_api: webrender_api::RenderApi,
+    /// Map of live WebGLContexts.
+    contexts: HashMap<WebGLContextId, GLContextWrapper>,
+    /// Cached information for WebGLContexts.
+    cached_context_info: HashMap<WebGLContextId, WebGLContextInfo>,
+    /// Current bound context.
+    current_bound_webgl_context_id: Option<WebGLContextId>,
+    /// Id generator for new WebGLContexts.
+    next_webgl_id: usize,
+    /// Handler user to send WebVR commands.
+    webvr_compositor: Option<VR>,
+    /// Generic observer that listens WebGLContext creation, resize or removal events.
+    observer: OB,
+}
+
+impl<VR: WebVRRenderHandler + 'static, OB: WebGLThreadObserver> WebGLThread<VR, OB> {
+    pub fn new(gl_factory: Arc<GLContextFactory>,
+               webrender_api_sender: webrender_api::RenderApiSender,
+               webvr_compositor: Option<VR>,
+               observer: OB) -> Self {
+        WebGLThread {
+            gl_factory: gl_factory,
+            webrender_api: webrender_api_sender.create_api(),
+            contexts: HashMap::new(),
+            cached_context_info: HashMap::new(),
+            current_bound_webgl_context_id: None,
+            next_webgl_id: 0,
+            webvr_compositor,
+            observer: observer,
+        }
+    }
+
+    /// Creates a new `WebGLThread` and returns a Sender to
+    /// communicate with it.
+    pub fn start(gl_factory: Arc<GLContextFactory>,
+                 webrender_api_sender: webrender_api::RenderApiSender,
+                 webvr_compositor: Option<VR>,
+                 observer: OB)
+                 -> WebGLSender<WebGLMsg> {
+        let (sender, receiver) = webgl_channel::<WebGLMsg>().unwrap();
+        let result = sender.clone();
+        thread::Builder::new().name("WebGLThread".to_owned()).spawn(move || {
+            let mut renderer = WebGLThread::new(gl_factory,
+                                                webrender_api_sender,
+                                                webvr_compositor,
+                                                observer);
+            let webgl_chan = WebGLChan(sender);
+            loop {
+                let msg = receiver.recv().unwrap();
+                let exit = renderer.handle_msg(msg, &webgl_chan);
+                if exit {
+                    return;
+                }
+            }
+        }).expect("Thread spawning failed");
+
+        result
+    }
+
+    /// Handles a generic WebGLMsg message
+    #[inline]
+    pub fn handle_msg(&mut self, msg: WebGLMsg, webgl_chan: &WebGLChan) -> bool {
```

Comment:
Yes, you are right. It was pub because I used it in the "packaging mode" I mentioned in the mailing list. But it can be private in the current implementation ;)